packages/server/lib/plugins/preprocessor.ts

```diff
@@ -0,0 +1,175 @@
+    ipc.send('preprocessor:close', filePath)
+  })
+})
+
+// for simpler stubbing from unit tests
+interface PreprocessorAPI {
+  errorMessage: (err?: PreprocessorError) => string
+  clientSideError: (err: PreprocessorError) => string
+  emitter: EventEmitter
+  getFile: (filePath: string, config: Cfg) => Promise<string>
+  removeFile: (filePath: string, config: Cfg) => void
+  close: () => void
+}
+
+const API: PreprocessorAPI = {
+  errorMessage,
+
+  clientSideError,
+
+  emitter: baseEmitter,
+
+  getFile (filePath: string, config: Cfg) {
+    let fileObject: FileObject
+
+    debugFn(`getting file ${filePath}`)
+    filePath = path.resolve(config.projectRoot, filePath)
+
+    debugFn(`getFile ${filePath}`)
+
+    if (!(fileObject = fileObjects[filePath])) {
+      // we should be watching the file if we are NOT
+      // in a text terminal aka cypress run
+      // TODO: rename this to config.isRunMode
+      // vs config.isInteractiveMode
+      const shouldWatch = !config.isTextTerminal || Boolean(process.env.CYPRESS_INTERNAL_FORCE_FILEWATCH)
+
+      const baseFilePath = filePath.replace(config.projectRoot, '')
+
+      fileObject = (fileObjects[filePath] = _.extend(new EventEmitter(), {
+        filePath,
+        shouldWatch,
+        outputPath: appData.getBundledFilePath(config.projectRoot, baseFilePath),
+      }))
+
+      fileObject.on('rerun', () => {
+        debugFn('file object rerun event')
+
+        return baseEmitter.emit('file:updated', filePath)
+      })
+
+      baseEmitter.once('close', () => {
+        debugFn('base emitter native close event')
+
+        return fileObject.emit('close')
+      })
+    }
+
+    // Check if we already have a processor for this file in headless mode
+    if (config.isTextTerminal && filePath in fileProcessors) {
+      debugFn('headless and already processed')
+
+      return fileProcessors[filePath]
+    }
+
+    const preprocessor = (fileProcessors[filePath] = Promise.try(() => {
+      const span = telemetry.startSpan({ name: 'file:preprocessor' })
+
+      return plugins.execute('file:preprocessor', fileObject).then((arg: string) => {
+        span?.setAttribute('file', arg)
+        span?.end()
+
+        return arg
+      })
+    }))
+
+    return preprocessor
+  },
+
+  removeFile (filePath: string, config: Cfg) {
+    let fileObject: FileObject
+
+    filePath = path.resolve(config.projectRoot, filePath)
+
+    if (!(filePath in fileProcessors)) {
+      return
+    }
+
+    debugFn(`removeFile ${filePath}`)
+
+    baseEmitter.emit('close', filePath)
+
+    fileObject = fileObjects[filePath]
+
+    if (fileObject) {
+      fileObject.emit('close')
+    }
+
+    delete fileObjects[filePath]
+
+    delete fileProcessors[filePath]
+  },
+
+  close () {
+    debugFn('close preprocessor')
+
+    fileObjects = {}
+    fileProcessors = {}
+    baseEmitter.emit('close')
+
+    baseEmitter.removeAllListeners()
+  },
+}
+
+export default API
+
+module.exports = API
```

Comment:
should be able to do the same here too. otherwise the compiled output could conflict